### ISSUE TITLE
ReactiveTools generate many model instances

### ISSUE THREAD

**Issue by hhaensel** [GenieFramework/Stipple.jl#152]
Currently, each call to `@in`, `@out` etc. generate a new model type. Generating many models is quite a waste of space, because types cannot be deleted.

The reason is that `Stipple.Pages._pages` are automatically updated:
```julia
  clear_type(__module__)
  instance = @eval __module__ @type()
  for p in Stipple.Pages._pages
    p.context == __module__ && (p.model = instance)
  end
```
Otherwise we could eliminate `instance = @eval __module__ @type()` and the model type would only be generated when calling `@model` or `@type`.
@essenciary I am not familiar with `_pages` and when they are called. Can we move the
```julia
  for p in Stipple.Pages._pages
    p.context == __module__ && (p.model = instance)
  end
```
to a later point?

**Comment by hhaensel** [comment]
Otherwise I propose two possible workarounds:
- add a block syntax for macros, e.g.
    ```julia
    @in begin
        a = 12
        s::AbstractString = "Why?"
    end
    ```
    That would reduce the number of definitions to maximum  (`@in`, `@out`, `@readonly`, `@private`, `@jsfn`)
- add ablock syntax like
    ```julia
    @bindings begin
        in:  a::Real = 12
        out: b::AbstractString = "hi"
    end
    ```

**Comment by essenciary** [comment]
Thanks! 

My take is: 
* we should improve and optimize continuously
* in general developer productivity comes at a cost. A higher level, more productive API is usually correlated to certain sacrifices in performance as increased time/space complexity. The key is to find the right balance (and usually this depends on the user base that the API targets). 

As such: 
* is the cost of the intermediate models really significant? We should try to get some numbers here - what is the avg size of a temporary type/instance (trying to figure out how to measure)
* manually calling `@model` or `@type` is the kind of tradeoff I mentioned above (in this case sacrificing dev productivity for reducing space/time complexity) - is it worth doing?
* as a principle I fully agree to adding non-breaking extra APIs to support more performant use cases - like `@bindings` and/or having a `@page!` macro that requires manually calling `@model` or `@type`.

**Comment by essenciary** [comment]
Overall, the ReactiveTools API still requires a lot of work, what we have now is just the first iteration. The `@page` macro could use some extra time and attention for sure. Also, a big issue is that we don't have a way to remove bindings from memory, once created.

**Comment by hhaensel** [comment]
>  Also, a big-ish issue is that we don't have a way to remove bindings from memory, once created.
We have `ReactiveTools.delete_bindings`.
I could imagine a macro `@clear` that either deleted all bindings or selected Symbols

**Comment by essenciary** [comment]
The issue is more about various way that the code can be run. 

When running the app in production (no more code editing, the app gets loaded non-interactively) there is no need to remove bindings as in theory the production code has only the necessary bindings. 

The issue though is when doing interactive work, for example by evaling lines of code in VSCode. Let's say we added some bindings, then we simply delete one (we remove the `@in` or `@out` macro in front of the variable). Do I now need to type and run `@clear` in the file/REPL? 

In theory it should be possible to detect when the `@in` or `@out` macro is removed (through Revise) and clear automatically?

**Comment by essenciary** [comment]
Actually even in production it might be interesting to change/remove bindings dynamically. For ex somebody might want to make a binding readonly after being writeable, based on some logic.

**Comment by hhaensel** [comment]
I just pushed `@clear` to the PR 😄 , usage
```julia
@clear
```
or
```julia
@clear x y z
```

**Comment by hhaensel** [comment]
From my end the refactoring is complete now.
@essenciary are you ok with eliminating variable definition from the macros as I have commented above?

In that case, I'd ask you for review ...

**Comment by hhaensel** [comment]
> Actually even in production it might be interesting to change/remove bindings dynamically. For ex somebody might want to make a binding readonly after being writeable, based on some logic.

Very true, Vue-people do that all the time - ... and they have 100 models on one page 🤣

**Comment by hhaensel** [comment]
with automatic type ...
```julia
@private non_reactive x = 12
# x__::Int64 = 32
```

**Comment by essenciary** [comment]
@hhaensel I'm not sure about converting `x` to `x__` - why don't we just add it to the internal reactive model as a non reactive field of the same name?

**Comment by essenciary** [comment]
@hhaensel Are these `x` -> `x__` fields handled in the `@handler` block? (I looked over the code but can't spot the logic).

**Comment by hhaensel** [comment]
Very true!
Actually, I think, it would be good to go away from the pattern matching. I'd propose to have two model fields
```julia
_private::Vector{Symbol} = Symbol[]
_readonly::Vector{ymbol} = Symbol[]
```
which hold the symbols of private and readonly non-reactive fields. We would only have to change some definitions like
```julia
isprivate(fieldname::Symbol, model::M) = fieldname in model._private
```

**Comment by essenciary** [comment]
Makes sense. But only for non-reactive fields, right? For reactive ones we have the properties and those work great.

**Comment by hhaensel** [comment]
Right!

**Comment by hhaensel** [comment]
Concerning naming: should we keep appending `__`? A Julia convention is to hide things with `_`-prefixing.

**Comment by hhaensel** [comment]
We chose it in order to align with javascript, which doesn't support leading `_`.

**Comment by essenciary** [comment]
`channels__` is an internal field that should not be used/touched by users (we have an API for getting/setting the channel). As such, I went with the postfix `__` to allow users to pick more common names (like `_channel`). 

JS does support leading `_`. 
<img width="98" alt="image" src="https://user-images.githubusercontent.com/11292378/202163198-7b463ed1-8f14-4d52-8b59-9bddd4ee9d8e.png">

**Comment by hhaensel** [comment]
But model fields of Vue models should not start with _ if I remember correctly

**Comment by essenciary** [comment]
Ah ok, I was not aware of that!

**Comment by hhaensel** [comment]
`channel__` is hardcoded in `@reactive!`
There is no easy way of getting rid of it. In all other places it's replaced.

Just pushed all changes. Please have a test 😃

**Comment by hhaensel** [comment]
Currently in the hh-abstract-models branch

**Comment by hhaensel** [comment]
This is solved now by #151
